Composer/packages/form-dialogs/src/undo/UndoRoot.tsx

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import {
+  Loadable,
+  RecoilState,
+  Snapshot,
+  useGotoRecoilSnapshot,
+  useRecoilSnapshot,
+  // eslint-disable-next-line @typescript-eslint/camelcase
+  useRecoilTransactionObserver_UNSTABLE,
+} from 'recoil';
+import * as React from 'react';
+import { AtomMap, History } from 'src/undo/types';
+
+const getAtomMap = (snapshot: Snapshot, trackedAtoms: RecoilState<any>[]): AtomMap => {
+  const atomMap = new Map<RecoilState<any>, Loadable<any>>();
+  for (const atom of trackedAtoms) {
+    atomMap.set(atom, snapshot.getLoadable(atom));
+  }
+  return atomMap;
+};
+
+const didAtomMapsChange = (prevMap: AtomMap, currMap: AtomMap) => {
+  if (prevMap.size !== currMap.size) {
+    return true;
+  }
+
+  for (const key of prevMap.keys()) {
```

Comment:
Is it possible for the sizes to be the same, but then currMap has a key that prevMap does not?  You could use lodash intersection/union methods to determine if there are any key differences.

Reply:
Good catch, I added that there when the trackedAtom size was fixed.